google/cloud/bigquery/v2/minimal/internal/dataset.h

```diff
@@ -0,0 +1,195 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_BIGQUERY_V2_MINIMAL_INTERNAL_DATASET_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_BIGQUERY_V2_MINIMAL_INTERNAL_DATASET_H
+
+#include "google/cloud/version.h"
+#include <nlohmann/json.hpp>
+#include <chrono>
+#include <memory>
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace bigquery_v2_minimal_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+// Disabling clang-tidy here as the namespace is needed for using the
+// NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT.
+using namespace nlohmann::literals;  // NOLINT
+
+auto const kDefaultTimepoint =
+    std::chrono::system_clock::time_point{std::chrono::milliseconds(0)};
+
+struct StorageBillingModel {
+  static StorageBillingModel UnSpecified();
+  static StorageBillingModel Logical();
+  static StorageBillingModel Physical();
+
+  std::string value;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(StorageBillingModel, value);
+
+struct TargetType {
+  static TargetType UnSpecified();
+  static TargetType Views();
+  static TargetType Routines();
+
+  std::string value;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(TargetType, value);
+
+struct TableFieldSchemaRoundingMode {
+  static TableFieldSchemaRoundingMode UnSpecified();
+  static TableFieldSchemaRoundingMode RoundHalfAwayFromZero();
+  static TableFieldSchemaRoundingMode RoundHalfEven();
+
+  std::string value;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(TableFieldSchemaRoundingMode,
+                                                value);
+
+struct DatasetReference {
+  std::string dataset_id;
+  std::string project_id;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(DatasetReference, dataset_id,
+                                                project_id);
+
+struct LinkedDatasetSource {
+  DatasetReference source_dataset;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(LinkedDatasetSource,
+                                                source_dataset);
+
+struct TableReference {
+  std::string project_id;
+  std::string dataset_id;
+  std::string table_id;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(TableReference, dataset_id,
+                                                project_id, table_id);
+
+struct RoutineReference {
+  std::string project_id;
+  std::string dataset_id;
+  std::string routine_id;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(RoutineReference, dataset_id,
+                                                project_id, routine_id);
+
+struct DatasetAccessEntry {
+  DatasetReference dataset;
+  std::vector<TargetType> target_types;
+};
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT(DatasetAccessEntry, dataset,
+                                                target_types);
+
+struct Access {
+  bool operator==(Access const& other);
```

Comment:
If you are going to define `operator==()` then you should define `operator!=` (it is easy: `inline bool operator!=(Access const& a, Access const& b) { return !(a == b); }`)

If this is going to be a member function (as you have it here) it should be `const`.

I think some folks recommend using a standalone function outside the class because conversion rules for the first argument work better with them.

Reply:
I do not see an implementation for this function.  Maybe just remove it?